app/lib/import_export.rb

```diff
@@ -600,90 +332,34 @@ def extract_patients_details(patients, fields)
-                  end
-        ]
-  end
-
-  # Gets a hash of the 2 latest laboratories of each patient
-  def get_patients_laboratories(patients)
-    Laboratory.where(patient_id: patients.pluck(:id)).order(specimen_collection: :desc).group_by(&:patient_id).transform_values { |v| v.take(2) }
-  end
-
-  # Extracts assessment data values given relevant fields
+  # Extract assessment data values given relevant fields
   def extract_assessments_details(patients_identifiers, assessments, fields)
     if fields.include?(:symptoms)
-      conditions = ReportedCondition.where(assessment_id: assessments.pluck(:id))
-      symptoms = Symptom.where(condition_id: conditions&.pluck(:id)).order(:label)
-
-      conditions_hash = Hash[conditions&.pluck(:id, :assessment_id)&.map { |id, assessment_id| [id, assessment_id] }]
+      conditions_hash = Hash[assessments.joins(:reported_condition).pluck('conditions.id', :assessment_id).map { |id, assessment_id| [id, assessment_id] }]
                         .transform_values { |assessment_id| { assessment_id: assessment_id, symptoms: {} } }
-      symptoms&.each do |symptom|
-        conditions_hash[symptom[:condition_id]][:symptoms][symptom[:name]] = symptom.value
+
+      # Uses index_symptoms_on_condition_id
+      all_symptoms = Symptom.where(condition_id: conditions_hash.keys)
+      all_symptoms.where(type: 'BoolSymptom').pluck(:condition_id, :name, :bool_value).each do |(condition_id, name, bool_value)|
+        conditions_hash[condition_id][:symptoms][name] = bool_value
+      end
+      all_symptoms.where(type: 'IntegerSymptom').pluck(:condition_id, :name, :int_value).each do |(condition_id, name, int_value)|
+        conditions_hash[condition_id][:symptoms][name] = int_value
+      end
+      all_symptoms.where(type: 'FloatSymptom').pluck(:condition_id, :name, :float_value).each do |(condition_id, name, float_value)|
+        conditions_hash[condition_id][:symptoms][name] = float_value
       end
       assessments_hash = Hash[conditions_hash.map { |_, condition| [condition[:assessment_id], condition[:symptoms]] }]
     end
 
-    symptom_names_and_labels = symptoms&.distinct&.pluck(:name, :label)
-
     assessments_details = []
     assessments.each do |assessment|
       assessment_details = assessment.custom_details(fields, patients_identifiers[assessment.patient_id]) || {}
-      if fields.include?(:symptoms)
-        symptom_names_and_labels&.map(&:first)&.each do |symptom_name|
-          # Nil check in case for some reason there were assessments with nil ReportedCondition
-          next if assessments_hash[assessment[:id]].nil?
 
+      # Nil check in case for some reason there were assessments with nil ReportedCondition
+      if fields.include?(:symptoms) && assessments_hash[assessment[:id]].present?
+        # Using Symptom.all for finding distinct symptom names is actually faster here because it can use symptoms_index_chain_1
+        Symptom.all.distinct.pluck(:name).each do |symptom_name|
```

Comment:
Same comment about using Symptom.all here. Also given that we make this same query twice in the lifetime of a single export, I'd be interested in seeing if we can consolidate that down to 1.

Reply:
I think this one actually doesn't matter whereas the other one does because even if a symptom not included in this jurisdiction's symptoms comes up in the loop, it'll just set `assessment_details[<symptom_name>]' to `nil` and that entry of the hash will not be touched when writing to file because only fields in `field_data` will be retrieved.

I think the performance gains of using Symptom.all over any kind of scoped or joined query far outweighs any other solution I can think of since this query takes milliseconds to run compared to ~30 seconds for the joined query.

As for consolidating the queries, I originally didn't think that was necessary since it only takes milliseconds to run, but I guess that can add up over tens and hundreds of batches.

Reply:
So I think I've figured out a decent solution to handle this so that we're getting symptom names from the patients selected for export while keeping the queries reasonable: https://github.com/SaraAlert/SaraAlert/pull/717/commits/ab0472956fb88101b83558e527cd1558fd81a1c0#diff-c34521fcd0fdc2034d456886aae8fd35299fe908ab88d8e49c60d0b4d39ab26bR193

It also now returns the `symptom_names` to be used later on in `extract_assessment_details`

Reply:
I actually ended up reimplementing `extract_assessment_details` in a way that doesn't require `symptom_names`. It's much more performant now because I was able to get rid of some intermediate hashes and arrays and reduce the number of queries: https://github.com/SaraAlert/SaraAlert/blob/35627a7fc3b92ac2f4466716feff0c76d9c006f1/app/lib/import_export.rb#L330